### PR TITLE
EDR 7: Design phase documentation (current result)

### PR DESCRIPTION
**Pull request by Dam-CZ**
Results of all the workshops conducted from April until September regarding problem framing, solution architecture, bounded context allocation

### CODE REVIEW

**Review by AndrewWestberg — APPROVED** [review]

File: engineering-decision-records/007-design-phase-documentation.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+The most discussed problem during our sessions was about the resources it takes to run a Cardano node today, this is currently a concern for most SPOs and we will need to solve that by making it the least demanding we can. 
+This also goes with the fact that current running a node on an ARM setup isn't easy, we want to solve that by being able to run the node with low cost, low consumption, low heat generation setups.
+
+Then the aspect of contribution, experimenting and customizing the node became the next topic to be discussed for people that currently run a node and the ones that build on top of it. The main concern was targeted towards having a closer look of the node inner-working to modify, select only the required setup to run for any use case. The modular approach of our architecture is directly linked to these concerns and will be one of our main decision making criteria.
+
+### People impacted
+
+Next on the menu was identifying who are the people that are in the environment of Amaru.
+
+<img src="https://i.imgur.com/z2uBl7F.png"
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md:51]:
```suggestion
<img src="https://github.com/user-attachments/assets/4db86523-6edd-4d8c-b590-c42b27f62539"
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 45%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Problem statements](https://github.com/user-attachments/assets/4db86523-6edd-4d8c-b590-c42b27f62539)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+### Desired outcomes: People
+
+For this category we asked ourselves what success would mean for people using Amaru, and the outcome was comforting the vision we had. 
+
+This will serve to draft out metrics that will drive our decision making towards meeting those success criteria.
+
+<img src="https://i.imgur.com/Fggxenr.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 45%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![People desired outcomes](https://github.com/user-attachments/assets/960e2e5f-1289-4722-b4dd-796ae486542a)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+### Desired outcomes: Business
+
+Now we looked at what success meant for the current market in which the Cardano node is.
+
+As mentioned above this will provide valuable data to be gathered to identify if Amaru is going the right direction in its market.
+
+<img src="https://i.imgur.com/dbsRYpU.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 50%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md:85]: Business desired outcomes

Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Business desired outcomes](https://github.com/user-attachments/assets/240172b1-a6de-4008-a810-0f460ef31448)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+### Constraints and risks
+
+As our final step before diving into the architecture of the solution we looked at the things we'll have to integrate as constraints and the potential risks that might come up on our project.
+
+The outcome of this will drive our synchronisation steps and the evolutions we plan for Amaru.
+
+<img src="https://i.imgur.com/PQg243u.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 80%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Constraints analysis](https://github.com/user-attachments/assets/33996d90-a677-4298-a08e-5f7ab6d1913c)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+> Rust focused peer reviews included in the development process 
+> 
+> Code audit and security audit
+
+To finalise our problem framing we ended up starting to draft risks that might impact the project.
+
+<img src="https://i.imgur.com/K2kcwiZ.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 80%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Risk Analysis](https://github.com/user-attachments/assets/67c6a3fe-4a8b-4bb2-95a3-e719c22ce7af)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+> Alignment on the Cardano Node next milestones and architecture modifications (working groups)
+
+### Architecture C4 diagram
+
+The next step towards building the product architecture was to use the [C4 Architecture](https://miro.com/app/board/uXjVNpawiPE=/?fromEmbed=1)
+
+<img src="https://i.imgur.com/zLGFoTr.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 50%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Context](https://github.com/user-attachments/assets/d55a8c9c-e16f-43fe-93b2-3d634c58c20e)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+- Upstream nodes
+- Downstream nodes
+- Client Apps
+
+Now if we dive a bit deeper into the system for Level 2, let's look at the containers and data stores inside our node:
+
+<img src="https://i.imgur.com/qUj8CMv.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 90%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Containers view](https://github.com/user-attachments/assets/c826ab17-c8fb-41e4-980b-8bdff60fdf27)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+Final step that we went through with the C4 model is the level 3: diving into the components inside each container.
+
+Consensus container:
+<img src="https://i.imgur.com/Pw9vdff.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 90%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Consensus container](https://github.com/user-attachments/assets/fbe9a00d-5cf6-4e03-8016-a3f578013b3a)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+Peer 2 Peer container:
+<img src="https://i.imgur.com/iKB2vtS.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 90%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Network container](https://github.com/user-attachments/assets/47cb1c15-55e7-4561-a35c-6ce2d8b28af3)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+RPC container:
+<img src="https://i.imgur.com/0IhAkCO.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 90%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![RPC Container](https://github.com/user-attachments/assets/fc52c608-ce39-4618-b7ca-edc1f10e7644)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+
+Transition container:
+<img src="https://i.imgur.com/0veQlnk.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 90%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Transition container](https://github.com/user-attachments/assets/23642f86-522c-4ba6-a9b5-283252563940)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+### Bounded context alignment
+
+The final design step was turning this architecture of the product into actionable parts that can be owned by a dedicated development team. 
+
+While thinking about that representation we used the bounded context methodology of Domain Driven Design to represent relationships, shared kernels and interfaces that will have to be monitored by each team that owns the bounded contexts. 
+
+<img src="https://i.imgur.com/kYr4ofP.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 100%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Bounded contexts](https://github.com/user-attachments/assets/3a1d3b6d-ff2f-44ba-8f4f-c3fcf4d722bd)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+- CF (Conformist): the downstream team must accept and adapt to the upstream team’s decisions
+- OHS (Open-Host Service): the supplier decouples its implementation logic from its public API to better serve consumers (can be subject to multiple integrations)
+- PL (Published Language): part of the domain that is exposed by the upstream member
+
+This fuelled our discussion on which part to dedicate a team for our project and we came up with the following independent teams (that might be subject to change as the project evolves):
+
+<img src="https://i.imgur.com/idFRxrD.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 100%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Maintainers](https://github.com/user-attachments/assets/6f27ba9e-fa87-4cfa-b03e-65efb8a38f91)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+- Cardano Ledger owner: Matthias Benkort
+- Consensus owner: Arnaud Bailly
+- Dolos owner: Santiago Carmuega
+- Forge owner: (*to be allocated*)
+- GO Node owner: Chris Gianelloni
+- Mempool owner: Andrew Westberg
+- Management RPC: (*to be allocated*)
+- Peer 2 peer (P2P) owner: Santiago Carmuega
+- Testnet facilitator owner: Chris Gianelloni
+- UTxO RPC specifications owner: Chris Gianelloni
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
- UTxO RPC specifications owner: Chris Gianelloni & Santiago Carmuega
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+This fuelled our discussion on which part to dedicate a team for our project and we came up with the following independent teams (that might be subject to change as the project evolves):
+
+<img src="https://i.imgur.com/idFRxrD.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 100%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
+
+For each scope identified, we nominated an owner that has the responsibility of managing the interfaces and the coherence of its scope:
+- Amaru integration: Amaru Maintainers Committee (*to be re-allocated*)
+- Cardano Ledger owner: Matthias Benkort
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
- Ledger owner: Matthias Benkort
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+The roles included were also of a varied range: CEO, COO, CTO, Senior developer, Project manager, SPO. 
+
+### Problem statements
+
+The first task at hand was formalising "what problems are we trying to solve by starting Amaru?"
+
+<img src="https://i.imgur.com/5NLPQ7t.png"
+     alt="sample image"
+     style="display: block; margin-right: auto; margin-left: auto; width: 45%;
+     box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)" />
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![Problem statements](https://github.com/user-attachments/assets/ae21dfbf-5f1a-4bac-8ca7-8b6603680546)
```

File: engineering-decision-records/007-design-phase-documentation.md [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+The most discussed problem during our sessions was about the resources it takes to run a Cardano node today, this is currently a concern for most SPOs and we will need to solve that by making it the least demanding we can. 
+This also goes with the fact that current running a node on an ARM setup isn't easy, we want to solve that by being able to run the node with low cost, low consumption, low heat generation setups.
+
+Then the aspect of contribution, experimenting and customizing the node became the next topic to be discussed for people that currently run a node and the ones that build on top of it. The main concern was targeted towards having a closer look of the node inner-working to modify, select only the required setup to run for any use case. The modular approach of our architecture is directly linked to these concerns and will be one of our main decision making criteria.
+
+### People impacted
+
+Next on the menu was identifying who are the people that are in the environment of Amaru.
+
+![Problem statements](https://github.com/user-attachments/assets/4db86523-6edd-4d8c-b590-c42b27f62539)
```
Comment (KtorZ) [inline engineering-decision-records/007-design-phase-documentation.md]:
```suggestion
![People impacted](https://github.com/user-attachments/assets/b23bee65-36a5-4660-ab63-0fc52cd25b96)
```

**Review by KtorZ — APPROVED** [review]

**Pull request merged**